File: 卡码/代码随想录/1/10-单调栈/Java/src/com/kurumi/LeetCode_0042.java
//42. ½ÓÓêË®
package com.kurumi;

import java.util.LinkedList;
import java.util.Stack;

public class LeetCode_0042 {

}
class Solution {
  public int trap(int[] height) {
  	int res=0;
  	LinkedList<Integer> st=new LinkedList();
  	for(int i=0;i<height.length;i++) {
  		while(!st.isEmpty()&&height[i]>height[st.peek()]) {
  			int mid=st.pop();
  			if(!st.isEmpty()) {
  				int h=Math.min(height[i],height[st.peek()])-height[mid];
  				int w=i-st.peek()-1;
  				res+=h*w;
  			}
  		}
  		st.push(i);
  	}
  	return res;
  }
}

File: 卡码/代码随想录/C++/单调栈/monotonicStack/LeetCode_0042.cpp
//42. 接雨水
#include <iostream>
#include <vector>
#include <stack>
using namespace std;

class Solution {
public:
	int trap(vector<int>& height) {
		stack<int> st;
		int res=0;
		for(int i=0;i<height.size();i++){
			while(!st.empty()&&height[i]>height[st.top()]){
				int mid=st.top();
				st.pop();
				if(!st.empty()){
					int h=min(height[i],height[st.top()])-height[mid];
					int w=i-st.top()-1;
					res+=h*w;
				}
			}
			st.push(i);
		}
		return res;
	}
};

int main(){
  system("pause");
  return 0;
}
